### DISCUSSION TITLE
Role of PipelineContext and PipelineContract

### DISCUSSION THREAD

**Discussion by Prateeksha-shanbhogue** [mtconnect/cppagent#369] [Q&A]
Hi there.!

I have been trying to understand the code flow..
Can anybody tell me how exactly does the pipelineContext and pipelineContract help in transforming the data from source to the sink....?
Thank you!

**Comment by wsobel** [comment]
They provide a narrow interface (contract) between the pipeline and the agent and various shared resources in the context to allow for counters, stats, dup detection, and the like. It is much easier to provide a controlled mocked environment than the entire MTConnect Agent to isolate testing of the transforms and pipelines to just the individual functionality. It is usually good practice to limit dependencies and separate concerns. 

The pipeline context provides a shared state between threads to manage some cross-data source accounting and management. Look at the period filter to get an idea of how the context works.

Best,
- Will Sobel
